File: ios/Methods/UserInteraction/PinEnrollMethodHandler.swift
//
// Copyright © 2023 Nevis Security AG. All rights reserved.
//

import NevisMobileAuthentication

struct PinEnrollMethodHandler: MethodHandler {
	func execute(using _: MobileAuthenticationClient, with message: ChannelInMessage) throws {
		let message: PinEnrollMessage = try validate(message: message)
		let operation: UserInteractionOperation = try OperationCache.shared.read(by: message.operationId)
		guard let state = operation.state as? EnrollPinState else {
			throw PluginError.illegalState("Enroll Pin state is expected!")
		}

		state.handler?.pin(message.pin)
		MethodChannelHandler.shared.resolve(method: .pinEnroll, message: nil)
	}
}

File: android/src/main/java/ch/nevis/mobile/authentication/sdk/react/methods/userInteraction/PinEnrollMethodHandler.kt
/*
 * Copyright © 2023 Nevis Security AG. All rights reserved.
 */

package ch.nevis.mobile.authentication.sdk.react.methods.userInteraction

import ch.nevis.mobile.authentication.sdk.react.MethodChannelHandler
import ch.nevis.mobile.authentication.sdk.react.cache.OperationCache
import ch.nevis.mobile.authentication.sdk.react.cache.operation.UserInteractionOperation
import ch.nevis.mobile.authentication.sdk.react.cache.state.EnrollPinState
import ch.nevis.mobile.authentication.sdk.react.methods.MethodHandler
import ch.nevis.mobile.authentication.sdk.react.model.ReactMethodName
import ch.nevis.mobile.authentication.sdk.react.model.message.incoming.ChannelInMessage
import ch.nevis.mobile.authentication.sdk.react.model.message.incoming.PinEnrollMessage
import ch.nevis.mobile.sdk.api.MobileAuthenticationClient

class PinEnrollMethodHandler : MethodHandler() {
    override fun execute(
        client: MobileAuthenticationClient,
        channelInMessage: ChannelInMessage
    ) {
        val message = validate<PinEnrollMessage>(channelInMessage)
        val operation = OperationCache.getOperation<UserInteractionOperation>(message.operationId)
        val currentState = operation.state as? EnrollPinState
            ?: throw IllegalStateException("Operation state is missing or not an enroll pin state!")
        currentState.enrollPin(message.pin)
        MethodChannelHandler.resolve(ReactMethodName.PIN_ENROLL, null)
    }
}
